pkg/downloader/chart_downloader.go

```diff
@@ -97,20 +97,29 @@ func (c *ChartDownloader) DownloadTo(ref, version, dest string) (string, *proven
 		return "", nil, err
 	}
 
-	data, err := g.Get(u.String(), c.Options...)
-	if err != nil {
-		return "", nil, err
-	}
-
 	name := filepath.Base(u.Path)
 	if u.Scheme == registry.OCIScheme {
 		idx := strings.LastIndexByte(name, ':')
 		name = fmt.Sprintf("%s-%s.tgz", name[:idx], name[idx+1:])
 	}
 
 	destfile := filepath.Join(dest, name)
-	if err := fileutil.AtomicWriteFile(destfile, data, 0644); err != nil {
-		return destfile, nil, err
+	_, err = os.Stat(destfile)
```

Comment:
There is a change in behavior here IMHO. If the chart has changed in the upstream registry, then the cache won't be invalidated.

It could be argued that released charts should never change (and I would tend to agree). The operative word is _should_, of course. Someone out there is probably updating a chart with re-versioning, and expecting it to work. I don't think we can break them here.

And prerelease version can be republished (under semver rules) I believe.

Reply:
the [digest](https://github.com/helm/helm/blob/8b4be22a81cb61e30affe3108abc84efd34b7f54/pkg/repo/index.go#L273) from the repo could be used to check wiether the chart needs to be downloaded again

Reply:
> There is a change in behavior here IMHO. If the chart has changed in the upstream registry, then the cache won't be invalidated.
> 
> It could be argued that released charts should never change (and I would tend to agree). The operative word is _should_, of course. Someone out there is probably updating a chart with re-versioning, and expecting it to work. I don't think we can break them here.
> 
> And prerelease version can be republished (under semver rules) I believe.

Yes, that is a possibility. Then fetching the chart metadata only also requires a network request. Perhaps we can use the --cache flag to specify whether to download or not?

Reply:
> There is a change in behavior here IMHO. If the chart has changed in the upstream registry, then the cache won't be invalidated.
> 
> It could be argued that released charts should never change (and I would tend to agree). The operative word is _should_, of course. Someone out there is probably updating a chart with re-versioning, and expecting it to work. I don't think we can break them here.
> 
> And prerelease version can be republished (under semver rules) I believe.

Yes, that is a possibility. Then fetching the chart metadata only also requires a network request. Perhaps we can use the --cache flag to specify whether to download or not?

Reply:
> Then fetching the chart metadata only also requires a network request. Perhaps we can use the --cache flag to specify whether to download or not?

In my world, the whole idea of local caching is to avoid network requests, i.e. if an application request can be satisfied locally it should do so (my particular use-case is doing a lot of `helm template --repo ...` locally, so I'm obviously biased). Re-released tarballs should not be the norm and there are ways to accomodate them without inducing network requests for well-behaving charts, i.e. re-fetch metadata only on pre-release charts, add a `--no-cache` option, etc. 

But this is still a (somewhat) breaking change. My personal preference would be to mention this in the release notes and leave it at that.